virtio-devices/src/watchdog.rs

```diff
@@ -0,0 +1,432 @@
+            }
+
+            used_desc_heads[used_count] = (avail_desc.index, len);
+            used_count += 1;
+        }
+
+        for &(desc_index, len) in &used_desc_heads[..used_count] {
+            queue.add_used(&mem, desc_index, len);
+        }
+        used_count > 0
+    }
+
+    fn signal_used_queue(&self) -> result::Result<(), DeviceError> {
+        self.interrupt_cb
+            .trigger(&VirtioInterruptType::Queue, Some(&self.queues[0]))
+            .map_err(|e| {
+                error!("Failed to signal used queue: {:?}", e);
+                DeviceError::FailedSignalingUsedQueue(e)
+            })
+    }
+
+    fn run(
+        &mut self,
+        paused: Arc<AtomicBool>,
+        paused_sync: Arc<Barrier>,
+    ) -> result::Result<(), EpollHelperError> {
+        let mut helper = EpollHelper::new(&self.kill_evt, &self.pause_evt)?;
+        helper.add_event(self.queue_evt.as_raw_fd(), QUEUE_AVAIL_EVENT)?;
+        helper.add_event(self.timer.as_raw_fd(), TIMER_EXPIRED_EVENT)?;
+        helper.run(paused, paused_sync, self)?;
+
+        Ok(())
+    }
+}
+
+impl EpollHelperHandler for WatchdogEpollHandler {
+    fn handle_event(&mut self, _helper: &mut EpollHelper, event: &epoll::Event) -> bool {
+        let ev_type = event.data as u16;
+        match ev_type {
+            QUEUE_AVAIL_EVENT => {
+                if let Err(e) = self.queue_evt.read() {
+                    error!("Failed to get queue event: {:?}", e);
+                    return true;
+                } else if self.process_queue() {
+                    if let Err(e) = self.signal_used_queue() {
+                        error!("Failed to signal used queue: {:?}", e);
+                        return true;
+                    }
+                }
+            }
+            TIMER_EXPIRED_EVENT => {
+                // When reading from the timerfd you get 8 bytes indicating
+                // the number of times this event has elapsed since the last read.
+                let mut buf = vec![0; 8];
+                if let Err(e) = self.timer.read_exact(&mut buf) {
+                    error!("Error reading from timer fd: {:}", e);
+                    return true;
+                }
+                if let Some(last_ping_time) = self.last_ping_time.lock().unwrap().as_ref() {
+                    let now = Instant::now();
+                    let gap = now.duration_since(*last_ping_time).as_secs();
+                    if gap > WATCHDOG_TIMEOUT {
+                        error!("Watchdog triggered: {} seconds since last ping", gap);
+                        self.reset_evt.write(1).ok();
+                    }
+                }
+                return false;
+            }
+            _ => {
+                error!("Unexpected event: {}", ev_type);
+                return true;
+            }
+        }
+        false
+    }
+}
+
+/// Virtio device for exposing entropy to the guest OS through virtio.
```

Comment:
I suppose this comment is a copy-paste from virtio-rng?

Reply:
Nice, thanks, fixed.